### PR TITLE
Source accession update

### LINKED ISSUE #30: Janno-Column for ENA Accession IDs
We will need a column in the Janno-Files for ENA-Accession numbers leading to alignment files. This will then enable - in principle - systematic reanalyses of individuals described in Poseidon packages and in the long run enable more uniform processing of genotypes described in package repositories.

### LINKED ISSUE #32: Data source/preparation pipeline
As discussed today we have to find a solution to preserve genotype data quality and document the preparation pipeline for each package. I opened this issue to discuss the concrete implementation.

Probably we should have an optional field in the POSEIDON.yml for that. Something like `genotypeDataPreparationProtocol`. We could limit this field to a hard selection of protocols (like `AARD` or `MPI-SHH-DAG`), where each of these values references a specific computational data preparation workflow starting with FASTQ or BAM-files. 

For `AARD` we only know what's written on the [website](https://reich.hms.harvard.edu/allen-ancient-dna-resource-aadr-downloadable-genotypes-present-day-and-ancient-dna-data): 
> The genotypes in the AADR are not always a perfect match to those from the associated published papers. This is because to make it easier to coanalyse datasets we have started from bam or fastq files, trimmed the ends of sequences to reduce errors due to ancient DNA damage in a way that is largely uniform across datasets, which may be slightly different from that used in the individual publications, and determined genotypes anew by sampling a random sequence to cover each position. 

For `MPI-SHH-DAG` (or a different name for "our" procedure) we should imho prepare a good documentation of the workflow - maybe with protocols.io (or or any other publication medium). The latter would indeed support computational workflows (e.g. [here](https://www.protocols.io/view/de-novo-transcriptome-assembly-workflow-ghebt3e)). Ideally we would have such a documentation also for the AARD, but that's not on us.

An additional question: What about modern data? It probably goes neither through the `AARD` nor the `MPI-SHH-DAG` pipeline.

Independent of this we could also store data source information in the POSEIDON.yml. This could even replace the individual-wise information we planned in #30. I imagine an optional list field `rawSequencingDataSource`, maybe like this:

```
rawSequencingDataSource:
- archive: ENA
  accession: BN000065
  url: https://www.ebi.ac.uk/ena/browser/view/BN000065
- archive: SRA
  accession: SRR000001
  url: https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=run_browser&run=SRR000001
- url: https://www.my_website.org/my_sequencing_data/1234
- archive: zenodo
  doi: 10.5281/zenodo.3346055
```
Each list entry has one or multiple of `archive`, `accession`, `url` or `doi`.

### LINKED ISSUE #33: SNP-Set specification in YAML file
There currently is no systematic way to query the specific SNP list of the genotype data. While we do have a column named "Data_Type" in the Janno file, which can be "Shotgun", "1240K", "OtherCapture" and so on, this doesn't imply the SNP set itself. For example, a sample can be "Shotgun" but still be represented by a 1240k SNP list.

It would be very useful to be able to check the SNP set upfront, because - for example - you might have a dataset with very low coverage and you know that you would want to squeeze out as much as possible, so would like to forge only packages with a full 1240K SNP list and not only Human Origins. I would suggest to add an optional field into the `genotypeData` object in the YAML file called `snpList` with restricted values, such as `HO`, `1240k` and in the future maybe even "1000G" if we allow genotype data to be of even denser format.

### LINKED ISSUE #34: The Publication_Status .janno column should support multiple entries
With a string list separated by `;`

### PR DESCRIPTION
**Pull request by stschiff**
This PR implements several or all of the outstanding issues #30 #32 #33

### CODE REVIEW

**Comment by stschiff** [comment]
Now this would also close #34

**Comment by stschiff** [comment]
One outstanding issue is whether want to make the `Genotype_SNP_set` a choice column. Currently this would make sense, as we have strictly only HumanOrigins and 1240K data. But we might want to include more datasets in the future, so I'm hesitant about this restriction.

**Comment by nevrome** [comment]
Ja - better not restrict that.

Generally your suggested solutions here have minimal complexity. Could be an issue down the road - or a big advantage. I think it's ok to go ahead with this now.

Regarding `Genetic_Source_Data`: I suggest to add a meaningful ordering criterion to the IDs. So instead of just 

> Accession ID(s) or URLs pointing to the source data used to generate the genotyping data.

> Accession ID(s) or URLs pointing to the source data used to generate the genotyping data. If multiple are given they should be arranged by descending specificity (e.g. project id > sample id > sequencing run id).

If this is merged we should not forget the following ToDos:

1. Add a more detailed description of these columns to https://poseidon-framework.github.io/#/janno_details
2. Update poseidonR to handle these new columns correctly
3. Update trident to handle these new columns correctly
4. (optional) Add these columns to the packages in our repo

Did I miss anything?

**Comment by stschiff** [comment]
Thanks. Possibilities to make more concrete would involve rules outside of our janno tsv file. For example, we could enforce `Data_Preparation_Pipeline` to be a valid URL, or we could enforce that the accession IDs follow the standard form on the ENA (or NCBI SRA?)

**Comment by stschiff** [comment]
I realised I was way too quick here. I have several suggestions which I'd like to commit, comments welcome:

**Genotype_SNP_set**: This is much more complicated than I thought. First of all, this is clearly a property of the package, not of the individual. So it shouldn't go into the Janno file, but into the package YAML file. You can quickly see that this is important when you think about forging two packages with different SNP sets. If you forge a 1240K with a HumanOrigins dataset, you will get a 1240K set if you merge by union, and HumanOrigins if you merge by intersect. I therefore suggest to add a "snp_set" variable to the `genotypeData` object in the YAML file, and strictly allow only
* `1240k`
* `HumanOrigins`
* `other`

and forge will follow simple rules to determine the snpset of a forge operation, depending on union or intersect settings and knowing that 1240K is a strict superset of HumanOrigins. I would actually make this a required field, not optional. It will be easy to add to all the packages we have, since all ancient datasets currently are 1240k, and the modern ones are 50/50 or so, which are quick to check.

**Data_Preparation_Pipeline**: I suggest to rename to `Data_Preparation_Pipeline_URL` and make it a strict requirement that this be a valid URL. Perhaps we can enforce that even in `trident` and `validate` and the like.

**Genetic_Source_Data**: I suggest to rename to `Genetic_Source_Accession_IDs` and make it a strict requirement that this be a list of valid accession IDs, perhaps even with the hierarchical ordering that Clemens suggested (although I would for now not check this). For ENA this is simple, because these IDs follow a strict naming scheme. We will have to google how that looks for the NCBI SRA.

**Comment by stschiff** [comment]
OK, I've updated the PR. Have a look at the changes, please.

**Comment by nevrome** [comment]
I'm still not convinced Genotype_SNP_set should be an information on the package level. The outcome of `union` vs. `intersect` should clearly be reflected on the individual level: A sample with HO coverage will never have any information outside of this SNP set, so `union` would leave it at HO, even if it has now 600k additional missing values in the dataset. So HO will always stay HO, but 1240K can become HO if merged with `intersect`.

Maybe you have some input, @AyGhal, @wolfgangaroo or @hringbauer.

**Comment by stschiff** [comment]
Ah, important discussion!

OK, first of all, let me just say that the specific use case I had in mind would definitely benefit from a package-level indicator about the SNP set used in the genotype matrix of the package. For example, you might want to predict the effect of a forge-intersect. Or you perhaps want to screen for the allele frequency of a specific gene and therefore want to first see whether a specific package even contains that SNPs in its genotype file.

Going more into detail, I think there are three independent levels related to SNP calling:
1) Whether the library has been captured on a SNP-set or not. -> This is individual-based and already in the Poseidon Janno file under `Data_Type` (which by the way is a terrible name). 
2) On which SNP set the original BAM file has been pulled down do -> We don't have that right now but that information would also be individual-based, as @nevrome indicated.
3) Which SNP set is _de facto_ listed in the genotype data of the package -> This is what I had in mind with this Pull Request and I think it would have to be package-level information, because it's strictly information about the genotype data files.

That these three levels are really independent is easy to see. For example, you might have a Shotgun library which has been pulled down to 1240K but listed only among Human-Origins SNPs in that particular package, due to a previous forge. Or you might have a 1240K-capture library, which was originally pulled down to 1000 Genomes and then imputed, but still listed in the SNP file only among 1240K SNPs.

What happens to these levels if you forge data? Arguably, Levels 1 and 2 won't change, but level 3 will. I can't right now think of a good use-case for why we need level 2, but I'm happy to consider that, too. But I think we would definitely benefit from a simple indicator on the package level about the row-shape (i.e. the SNP set) of the genotype matrix, as in the PR.

What do you think?

By the way, I realised that there is an issue with forge-intersect with the janno field `Nr_autosomal_SNPs`, which might change. Also, we might have to reconsider the `Coverage_1240k` column, which also seems hardcoding the 1240k SNP set, while in the future we might have different pull downs as well... I'll open separate issues for those.

**Comment by stschiff** [comment]
We had a brief meeting today (Apr 30) and decided we can in principle go ahead with this proposal of a new package-level field called `snpset` and the old `Data_type` janno field. @nevrome please have one more quick look, then we can merge. And I'll quickly go ahead and bump the version nr.

**Comment by stschiff** [comment]
Version nr will be done only after the merge, actually.

File: POSEIDON_yml_fields.tsv [hunk 1]
```diff
@@ -14,7 +14,8 @@ genoFileChkSum	1	genotypeData	md5 checksum of the genoFile	String		FALSE	TRUE
 snpFile	1	genotypeData	relative path to snpFile	String	Path	TRUE	TRUE
 snpFileChkSum	1	genotypeData	md5 checksum of the snpFile	String		FALSE	TRUE
 indFile	1	genotypeData	relative path to indFile	String	Path	TRUE	TRUE
-indFileChkSum	1	genotypeData	md5 checksum of the indFile	String		FALSE	TRUE
+indFileChSum	1	genotypeData	md5 checksum of the indFile	String		FALSE	TRUE
```
Comment (nevrome) [inline POSEIDON_yml_fields.tsv]: `indFileChSum` should be changed back to `indFileChkSum`. Typo?

**Review by nevrome — COMMENTED** [review]
See my one comment. Beyond that I don't see any issues (yet :sweat_smile:)

Can you prepare the respective downstream issues in the three affected repos?

**Comment by nevrome** [comment]
I can also open these issues - wait.

File: POSEIDON_yml_fields.tsv (hunk 1 above)
Comment (stschiff) [reply POSEIDON_yml_fields.tsv]: Oops, yes, of course. No idea how that happened.

Comment (stschiff) [reply POSEIDON_yml_fields.tsv]: Done.

**Comment by nevrome** [comment]
I compiled what came to my mind for the downstream issues. At least 16 individual ToDos :grimacing:

**Pull request merged**